packages/react-forms/src/InputDate/index.js

```diff
@@ -21,141 +22,177 @@ function toYYYYMMDD(date) {
-
-    const dateValue = parse(value);
-    const isDateValid = !isNaN(dateValue.getTime());
-
-    return (
-      <MouseOutside
-        onClickOutside={this.handleClose}
-        refs={[this.refA, this.refB]}
-      >
-        {() => (
-          <Manager>
-            <Reference>
-              {({ ref }) => (
-                <InputText
-                  type="date"
-                  renderAddon={({ marginRightClass }) => (
-                    <Icon
-                      name="calendar"
-                      className={marginRightClass}
-                      onClick={!props.disabled ? this.handleOpen : undefined}
-                    />
-                  )}
-                  onFocus={this.handleOpen}
-                  focus={isOpen}
-                  onBlur={this.handleCloseOnBlur}
-                  value={value}
-                  onChange={this.handleInputChange}
-                  onClick={this.preventDefault}
-                  min={min}
-                  max={max}
-                  ref={mergeRefs(ref, this.refA)}
-                  {...props}
-                />
-              )}
-            </Reference>
-            {isOpen && (
-              <Popper placement="bottom">
-                {({ ref, style }) => (
-                  <Calendar
-                    onChangeCurrent={current => this.setState({ current })}
-                    current={current || (isDateValid ? dateValue : new Date())}
-                    onSelect={this.handleSelect}
-                    selected={isDateValid ? dateValue : undefined}
-                    minDate={min ? new Date(min) : undefined}
-                    maxDate={max ? new Date(max) : undefined}
-                    ref={mergeRefs(ref, this.component, this.refB)}
-                    style={style}
+  }, [isOpen, setOpen]);
+
+  const handleCloseOnBlur = useCallback(
+    (evt: MouseEvent) => {
+      const { relatedTarget } = evt;
+      // istanbul ignore next (this is too DOM related)
+      if (
+        component.current && relatedTarget instanceof HTMLElement
+          ? !component.current.contains(relatedTarget)
+          : true
+      ) {
+        setOpen(false);
+      }
+    },
+    [component, setOpen]
+  );
+
+  const setCurrentOnValueChange = useCallback(
+    value => {
+      const current = new Date(value);
+      // istanbul ignore else
+      if (!isNaN(current.getTime())) {
+        setCurrent(current);
+      }
+    },
+    [setCurrent]
+  );
+
+  const handleInputChange = useCallback(
+    (evt: SyntheticInputEvent<HTMLInputElement>) => {
+      onChange(evt.target.value);
+    },
+    [onChange]
+  );
+
+  const handleSelect = useCallback(
+    (selected: Date) => {
+      setOpen(false);
+      onChange(toYYYYMMDD(selected));
+    },
+    [setOpen, onChange]
+  );
+
+  //Syncs calendar page with value
+  //Avoid first run for defaultCalendarValue to take effect
+  //Avoid when onCalendarChange or calendarValue is defined
+  useEffect(() => {
```

Comment:
The code below looks a lot imperative, I'd try to convert this into declarative to make it work better with the React functional programming patterns.

I don't understand why you are trying to keep track of this state internally instead of relying on the `useControlledState` which should handle this case for you.

`value`, `defaultValue`, and `onChange` are a group, `isOpen`, `defaultIsOpen`, and `onToggle` is another group, `defaultCalendarValue`, `calendarValue`, and `onCalendarChange` is another one.

All of them can be handled cleanly with `useControlledState`

Reply:
This is needed when value changes (for example user types another date in the input) I override the current page in the calendar selector. I need a way to keep track of this change. As far as `if (!(onCalendarChange || calendarValue)) {` If the `onCalendarChange` or `calendarValue` is defined we don't want to override the user defined value, because than it becomes the users responsibility to do that kind of change.
Keeping track of first run is to not change the calendar page when defaultCalendarValue is defined. It would change because `value` is given and it would automatically override it to the calendar page based on `value`.

Reply:
I don't think the code as-is is a good candidate to be merged, the current visualized calendar page should be tracked inside the state handled by the `useControlledState(defaultCalendarValue, calendarValue, onCalendarChange)`, and shouldn't be tracked elsewhere, so that the consumer is still able to manually control it if they want.

If you need help with this ping me on Slack please